# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project 2: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: **

This is a classification problem because students will be identified as "in need of early intervention" or not, which is a binary classification. This could be approached as a regression problem, where each student would be given a score (maybe from one to 100) to quantify their risk of failing to graduate. However, since the outcome of interest is binary (intervention or not) and a regression score would only be used to determine a binary outcome, a classification approach makes more sense.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [66]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [67]:
# TODO: Calculate number of students
n_students = student_data.shape[0]

# TODO: Calculate number of features
n_features = student_data.shape[1]

# TODO: Calculate passing students
n_passed = (student_data.loc[:,['passed']]=='yes').sum()[0]

# TODO: Calculate failing students
n_failed = (student_data.loc[:,['passed']]=='no').sum()[0]

# TODO: Calculate graduation rate
grad_rate = (student_data.loc[:,['passed']]=='yes').mean()[0]

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 31
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 0.67%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [68]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [69]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print X_all.head
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

<bound method DataFrame.head of      school_GP  school_MS  sex_F  sex_M  age  address_R  address_U  \
0          1.0        0.0    1.0    0.0   18        0.0        1.0   
1          1.0        0.0    1.0    0.0   17        0.0        1.0   
2          1.0        0.0    1.0    0.0   15        0.0        1.0   
3          1.0        0.0    1.0    0.0   15        0.0        1.0   
4          1.0        0.0    1.0    0.0   16        0.0        1.0   
5          1.0        0.0    0.0    1.0   16        0.0        1.0   
6          1.0        0.0    0.0    1.0   16        0.0        1.0   
7          1.0        0.0    1.0    0.0   17        0.0        1.0   
8          1.0        0.0    0.0    1.0   15        0.0        1.0   
9          1.0        0.0    0.0    1.0   15        0.0        1.0   
10         1.0        0.0    1.0    0.0   15        0.0        1.0   
11         1.0        0.0    1.0    0.0   15        0.0        1.0   
12         1.0        0.0    0.0    1.0   15        0.0   

### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [70]:
# TODO: Import any additional functionality you may need here

import numpy as np
from sklearn.cross_validation import train_test_split


# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=num_test, train_size=num_train, random_state=0)



# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **

## Support Vector Machines (SVM)

### Industry appliction: 
A study published in PNAS this year described the use of SVMs to predict the pathologic potential of different strains of E. coli isolated from cows using genomic data. /n
### Strengths: 
SVMs, when combined with the kernel trick, are great at modeling data that is not linearly separable. The kernel trick provides a wide range of functions to model complex relationships. /n
### Weaknesses: 
The use of the kernel trick to separate non-linear data can make it difficult to interpret the inner workings of the algorithm. In other words, SVMs can be somewhat of a black box, whereas other algorithms, like decision trees, are easier to interpret and to follow the reasoning behind a given classification. SVM could also be overkill if the underlying function turns out to have relatively low dimensionality. Training time can be much longer with SVMs if using non-linear kernels.
### Why SVM?: 
SVM would work well in this case because they offer more flexibility via kernels, and we don't have to worry about long training times because the data set is relatively small. 

## Decision Trees

### Industry application:
A recent paper described the use of decision trees to classify websites as 'malicious' or not.
### Strengths:
Decision trees implicitly perform feature selection, can capture non-linear relationships, require little data scrubbing, and they are easy to interpret, diagram, and explain. 
### Weaknesses:
They often overfit and they aren't 'smooth' (i.e. they don't draw smooth borders between data points, like SVM + kernels do).
### Why decision trees?:
This is likely a relatively simple problem that could be captured well by decision trees, and the result will be easy to interpret. 


## Gaussian Naive Bayes (GaussianNB)

### Industry application:
A recent paper in the journal of Clinical Neuroscience described the use of naive bayes to identify evidence of stroke on MRI scans.
### Strengths:
Simple. NB is great at accounting for 'blurry' lines in the data (i.e. data that is not easily separable linearly or non-linearly). 
### Weaknesses:
Can't learn interactions between features.
### Why NB?:
NB would be a simple and easy solution to the problem that could be easily implemented with a short run time.


### References:
* Lupolova, Nadejda, et al. "Support vector machine applied to predict the zoonotic potential of E. coli O157 cattle isolates." Proceedings of the National Academy of Sciences 113.40 (2016): 11312-11317.
* Mašetic, Zerina, Abdulhamit Subasi, and Jasmin Azemovic. "Malicious Web Sites Detection using C4. 5 Decision Tree." Southeast Europe Journal of Soft Computing 5.1 (2016).
* Griffis, Joseph C., Jane B. Allendorfer, and Jerzy P. Szaflarski. "Voxel-based Gaussian naïve Bayes classification of ischemic stroke lesions in individual T1-weighted MRI scans." Journal of neuroscience methods 257 (2016): 97-108.
* http://blog.echen.me/2011/04/27/choosing-a-machine-learning-classifier/
* http://www.yaksis.com/posts/why-use-svm.html
* http://www.simafore.com/blog/bid/62333/4-key-advantages-of-using-decision-trees-for-predictive-analytics




### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [71]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [72]:
# TODO: Import the three supervised learning models from sklearn

import numpy as np
from sklearn.svm import SVC
from sklearn import tree
from sklearn.naive_bayes import GaussianNB

# TODO: Initialize the three models
clf_A = SVC()
clf_B = tree.DecisionTreeClassifier()
clf_C = GaussianNB()

clfs = [clf_A, clf_B , clf_C]

# TODO: Set up the training set sizes
X_train_100 = X_train.iloc[0:100,]
y_train_100 = y_train.iloc[0:100,]

X_train_200 = X_train.iloc[0:200,]
y_train_200 = y_train.iloc[0:200,]

X_train_300 = X_train
y_train_300 = y_train

# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)
 
train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)

train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
    
train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)


Training a SVC using a training set size of 100. . .
Trained model in 0.0015 seconds
Made predictions in 0.0008 seconds.
F1 score for training set: 0.8591.
Made predictions in 0.0008 seconds.
F1 score for test set: 0.7838.
Training a SVC using a training set size of 200. . .
Trained model in 0.0038 seconds
Made predictions in 0.0027 seconds.
F1 score for training set: 0.8693.
Made predictions in 0.0013 seconds.
F1 score for test set: 0.7755.
Training a SVC using a training set size of 300. . .
Trained model in 0.0077 seconds
Made predictions in 0.0055 seconds.
F1 score for training set: 0.8692.
Made predictions in 0.0017 seconds.
F1 score for test set: 0.7586.
Training a DecisionTreeClassifier using a training set size of 100. . .
Trained model in 0.0009 seconds
Made predictions in 0.0002 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0002 seconds.
F1 score for test set: 0.6942.
Training a DecisionTreeClassifier using a training set size of 200. . .
Trained model in 

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - SVM **  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |  0.0014 seconds         |   0.0008      |   0.8591               |     0.7838            |
| 200               |      0.0035          |       0.0025          |    0.8693         |      0.7755       |
| 300               |      0.0079            |     0.0050             |   0.8692         |    0.7586      |

** Classifer 2 - Decision Tree **  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |      0.0014                   |   0.0002        |     1.0000      |     0.6491      |
| 200               |     0.0016             |      0.0002         |     1.0000       |     0.7328      |
| 300               |      0.0021           |       0.0003           |    1.0000       |     0.7458     |

** Classifer 3 - GaussianNB **  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |    0.0008            |      0.0003             |     0.8550      |     0.7481      |
| 200               |    0.0008               |      0.0004            |   0.8321         |   0.7132       |
| 300               |    0.0010               |      0.0007            |   0.8088         |   0.7500        |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: **

SVM performed the best. Even with relatively few data points (100) SVM outperformed the other two algorithms when using all of the available training data (300). SVM  took the longest to train and to predict (by an order of magnitude), but the actual run time difference is negligible. If this was intended for real-time prediction, run time could play a factor but in this scenario run time differences of this magnitude are completely irrelevent. 

Because SVM performed so well with limited data, future models may not require gather data from so many students (which could cut costs). Or efforts could be spent on collecting a greater variety of data points (like data related to sibling school performance for instance) rather than polling more individual students.

SVM also has the capabilities to capture interactions between parameters, which is more representative of the complex factors at play that can determine the success of a student.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **

Support vector machines work by finding hidden divisions within data. Imagine if you were trying to teach a computer to recognize clouds in the sky. SVM would approach this problem by looking at each part of the sky and drawing a line to divide blue from white. You can imagine this process happening in two dimensions or three dimensions. For a problem as complex as this one, we use dozens of dimensions but the problem is still the same. We are trying to separate students who will pass vs. those that won't, and SVM helps us find the borders between them.

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [77]:
# TODO: Import 'GridSearchCV' and 'make_scorer'

import numpy as np
import pandas as pd
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import fbeta_score, make_scorer

# TODO: Create the parameters list you wish to tune
C_range = np.logspace(-2, 10, 13)
gamma_range = np.logspace(-9, 3, 13)
param_grid = dict(gamma=gamma_range, C=C_range)

# TODO: Initialize the classifier
clf = SVC()

# TODO: Make an f1 scoring function using 'make_scorer' 

f1_scorer = make_scorer(fbeta_score, pos_label="yes", beta=2)

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method

grid_obj = GridSearchCV(clf, param_grid = param_grid, scoring=f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters

grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0054 seconds.
Tuned model has a training F1 score of 0.9717.
Made predictions in 0.0018 seconds.
Tuned model has a testing F1 score of 0.7919.


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **

The final F1 score is 0.9717 for training and 0.79 testing, both of which are an improvement from the previous models.

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.